### ISSUE TITLE
ttnn.grid_sample: the registered golden ignores use_precomputed_grid, returning NaN and a 3x-too-wide tensor for the documented precomputed path

### ISSUE THREAD

**Issue by p32929** [tenstorrent/tt-metal#57149]
### Summary

`ttnn.grid_sample` has two modes. The registered golden only implements one of them, and there is no error when it is handed the other — it silently returns **NaN and unrelated finite values, in a tensor 3x too wide**.

`golden_grid_sample` (`ttnn/ttnn/operations/pool.py:360`) never names `use_precomputed_grid`; the flag lands in `**_` (`pool.py:368`) and is discarded. The function then unconditionally does

```python
K_grid = last_dim // 2                                  # pool.py:397
grid_unpacked = grid.reshape(N, H_grid, total_W, 2)
torch.nn.functional.grid_sample(input_nchw, grid_unpacked, ...)
```

which only makes sense for a standard grid. A precomputed grid is `(N, H, W, 6*K)` and its 6 elements are **not coordinates** — per `grid_sample_prepare_grid.cpp:137-146` they are

| index | content |
|---|---|
| 0 | `h0` as an **int16 bit pattern** `bit_cast` into bfloat16 |
| 1 | `w0`, same |
| 2..5 | the four bilinear weights `nw, ne, sw, se`, as real bfloat16 |

So with `last_dim = 6` the golden derives `K_grid = 3` (the true K is 1), reshapes the 6 packed values into **three** coordinate pairs, and feeds bit patterns to `F.grid_sample` as normalized coordinates.

### Measured, on `4f9fa9e` (fp32 host, no device needed)

Input `(1, 4, 4, 2)`, grid `(1, 2, 3, 2)`, `align_corners=False`, `padding_mode="zeros"`:

```
CONTROL - device precomputed path vs the golden in standard mode
  golden(grid)           shape (1, 2, 3, 2)
  device(prepared_grid)  shape (1, 2, 3, 2)
  max abs diff           0.0436      (bf16 weight rounding only - weights are stored bf16 by design)

DEFECT - the same golden, handed the precomputed grid
  golden(prepared_grid)  shape (1, 2, 9, 2)   <-- 3x the width the device returns
  device                 shape (1, 2, 3, 2)

  K_grid the golden derives from last_dim=6 : 3   (true K is 1)
  first precomputed point read as floats    : [nan, nan, 0.0, 0.0, 0.0, 0.4902]
  golden reads pair 0 as (x, y) = (nan, nan)      <-- int16 bit patterns, not coordinates

  on the 3 overlapping columns:
    golden : [nan, nan, 15.0, 16.0, 22.844, 23.844, nan, nan]
    device : [0.0, 0.49, 19.398, 20.398, 24.258, 25.066, 0.0, 0.0]
    NaNs   : golden 4 of 12 elements, device 0
    max abs diff over elements finite in both: 9.41   (input values span 0..31)

  golden(prepared, use_precomputed_grid=True) == golden(prepared)  ->  True
                                                  (the flag is swallowed by **_)
```

The NaNs are not incidental. Any corner index that falls outside the image is stored as `h0 = -1`, i.e. `int16 0xFFFF`; read back as bfloat16 that bit pattern **is a NaN**, and `F.grid_sample` propagates it to the output. Every out-of-bounds sample — the normal case for warping and STNs, and exactly what `padding_mode="zeros"` exists to handle — turns into NaN in the reference.

### Why CI does not see this

`tests/ttnn/nightly/unit_tests/operations/pool/test_grid_sample.py:371` hands the golden the **original normalized grid** and gives only the device the prepared one:

```python
golden_grid_dtype = torch.float32 if use_precomputed_grid else torch.bfloat16
torch_output_nhwc = golden_grid_sample(
    input_tensor=torch_input_nhwc,
    grid=torch_grid.to(golden_grid_dtype),   # <-- the un-prepared grid
    ...
)
```

That is the correct expected value, and it is why the op itself is well tested — but it means the golden is never once called the way `ttnn.get_golden_function(ttnn.grid_sample)` invites a user to call it: with the same arguments that were passed to the op. The test works around the defect rather than exposing it.

### Also, a contradictory doc line

`grid_sample.hpp:27` says:

```
 *   use_precomputed_grid: Whether to use precomputed grid coordinates, currently only false is supported
```

but precomputed mode is fully implemented (`grid_sample_reader_common.hpp:58`, both program factories), is bound with its own default (`grid_sample_nanobind.cpp:145`), has a host-side helper (`ttnn.prepare_grid_sample_grid`) and is the documented performance path in the nanobind docstring, complete with a worked example at `grid_sample_nanobind.cpp:126-131`. One of the two statements is wrong; from the code it is the header comment.

### Reproducer

Device-free, stdlib + torch only, transcribes `prepare_grid_sample_grid`, the device reader's precomputed branch and `golden_grid_sample` verbatim, opens with a control row proving the transcriptions agree with the standard-mode golden to bf16 rounding, and **ends in asserts**.

<details>
<summary><code>grid_sample_precomputed_golden.py</code></summary>

```python
"""
ttnn.grid_sample: the registered golden ignores use_precomputed_grid.

Device-free. Pure torch. Transcribes, line for line:
  * ttnn.prepare_grid_sample_grid  (grid_sample_prepare_grid.cpp, bilinear + bfloat16 branch)
  * the device reader's precomputed path (grid_sample_reader_common.hpp,
    GridCoordinateReader<..., true>::read_grid_point + the validity mask)
  * ttnn.get_golden_function(ttnn.grid_sample)  (ttnn/ttnn/operations/pool.py, golden_grid_sample)

Run:  python3 grid_sample_precomputed_golden.py
"""
import torch

torch.manual_seed(0)

N, H_IN, W_IN, C = 1, 4, 4, 2
H_G, W_G = 2, 3

inp_nhwc = torch.arange(N * H_IN * W_IN * C, dtype=torch.float32).reshape(N, H_IN, W_IN, C)
inp_nchw = inp_nhwc.permute(0, 3, 1, 2).contiguous()

# normalized coords in [-1,1]; two of them deliberately out of bounds
grid = torch.tensor([[[[-0.9, -0.9], [0.1, 0.25], [0.8, 0.8]],
                      [[-1.6, 0.3], [0.0, 0.0], [0.45, -0.7]]]], dtype=torch.float32)
assert grid.shape == (N, H_G, W_G, 2)


def to_bf16(x):
    return torch.tensor(x, dtype=torch.float32).to(torch.bfloat16)


def int16_bits_as_bf16(v):
    """bit_cast<bfloat16>(uint16_t) — what prepare_grid_sample_grid does for h0/w0."""
    return torch.tensor([v], dtype=torch.int16).view(torch.bfloat16)[0]


def prepare_grid_sample_grid(grid, input_shape, align_corners=False):
    """grid_sample_prepare_grid.cpp, bilinear + OutputType=bfloat16. Returns (N,H,W,6) bf16."""
    input_h, input_w = input_shape[1], input_shape[2]
    if align_corners:
        h_scale = (input_h - 1) * 0.5 if input_h > 1 else 0.0
        w_scale = (input_w - 1) * 0.5 if input_w > 1 else 0.0
        h_off = w_off = 0.0
    else:
        h_scale, w_scale = input_h * 0.5, input_w * 0.5
        h_off = w_off = -0.5
    n, gh, gw, _ = grid.shape
    out = torch.zeros(n, gh, gw, 6, dtype=torch.bfloat16)
    for b in range(n):
        for h in range(gh):
            for w in range(gw):
                x_coord, y_coord = float(grid[b, h, w, 0]), float(grid[b, h, w, 1])
                h_img = (y_coord + 1.0) * h_scale + h_off
                w_img = (x_coord + 1.0) * w_scale + w_off
                h0, w0 = int(torch.floor(torch.tensor(h_img))), int(torch.floor(torch.tensor(w_img)))
                h1, w1 = h0 + 1, w0 + 1
                h0v, h1v = 0 <= h0 < input_h, 0 <= h1 < input_h
                w0v, w1v = 0 <= w0 < input_w, 0 <= w1 < input_w
                hf, wf = h_img - h0, w_img - w0
                hfi, wfi = 1.0 - hf, 1.0 - wf
                out[b, h, w, 0] = int16_bits_as_bf16(h0)
                out[b, h, w, 1] = int16_bits_as_bf16(w0)
                out[b, h, w, 2] = to_bf16(hfi * wfi if (h0v and w0v) else 0.0)
                out[b, h, w, 3] = to_bf16(hfi * wf if (h0v and w1v) else 0.0)
                out[b, h, w, 4] = to_bf16(hf * wfi if (h1v and w0v) else 0.0)
                out[b, h, w, 5] = to_bf16(hf * wf if (h1v and w1v) else 0.0)
    return out


def device_precomputed(inp_nhwc, prepared):
    """grid_sample_reader_common.hpp: read h0/w0 as int16 bits, 4 weights, mask invalid corners."""
    n, gh, gw, _ = prepared.shape
    _, ih, iw, c = inp_nhwc.shape
    out = torch.zeros(n, gh, gw, c, dtype=torch.float32)
    bits = prepared.view(torch.int16)
    for b in range(n):
        for h in range(gh):
            for w in range(gw):
                h0 = int(bits[b, h, w, 0]); w0 = int(bits[b, h, w, 1])
                h1, w1 = h0 + 1, w0 + 1
                wt = [float(prepared[b, h, w, 2 + k]) for k in range(4)]
                valid = [(0 <= h0 < ih and 0 <= w0 < iw), (0 <= h0 < ih and 0 <= w1 < iw),
                         (0 <= h1 < ih and 0 <= w0 < iw), (0 <= h1 < ih and 0 <= w1 < iw)]
                corners = [(h0, w0), (h0, w1), (h1, w0), (h1, w1)]
                acc = torch.zeros(c)
                for k in range(4):
                    if valid[k] and wt[k] != 0.0:
                        acc += wt[k] * inp_nhwc[b, corners[k][0], corners[k][1]]
                out[b, h, w] = acc
    return out


def golden_grid_sample(input_tensor, grid, mode="bilinear", padding_mode="zeros",
                       align_corners=False, batch_output_channels=False, grid_batching_factor=None, **_):
    """ttnn/ttnn/operations/pool.py — verbatim, including the **_ that swallows use_precomputed_grid."""
    n, h_grid, w_grid, last_dim = grid.shape
    k_grid = last_dim // 2
    input_nchw = input_tensor.permute(0, 3, 1, 2)
    c = input_nchw.shape[1]
    total_w = w_grid * k_grid
    grid_unpacked = grid.reshape(n, h_grid, total_w, 2)
    out_nchw = torch.nn.functional.grid_sample(
        input_nchw.float(), grid_unpacked.float(), mode=mode,
        padding_mode=padding_mode, align_corners=align_corners)
    out_nhwc = out_nchw.permute(0, 2, 3, 1).to(input_tensor.dtype)
    eff_k = grid_batching_factor if grid_batching_factor is not None else k_grid
    new_w = total_w // eff_k
    if batch_output_channels:
        out_nhwc = out_nhwc.view(n, h_grid, new_w, eff_k, c).contiguous().view(n, h_grid, new_w, c * eff_k)
    else:
        out_nhwc = out_nhwc.view(n, h_grid, new_w * eff_k, c)
    return out_nhwc


# ---------------------------------------------------------------- CONTROL ROW
ref = golden_grid_sample(inp_nhwc, grid)                       # standard mode, K=1
prepared = prepare_grid_sample_grid(grid, [N, H_IN, W_IN, C])  # ttnn.prepare_grid_sample_grid
dev = device_precomputed(inp_nhwc, prepared)                   # device, use_precomputed_grid=True

print("CONTROL — device precomputed path vs the golden in standard mode")
print("  golden(grid)              shape", tuple(ref.shape))
print("  device(prepared_grid)     shape", tuple(dev.shape))
print("  max abs diff             ", float((ref - dev).abs().max()),
      " (bf16 weight rounding only; weights are stored bf16 by design)")
assert ref.shape == dev.shape
assert float((ref - dev).abs().max()) < 0.2, "device model must track the standard-mode golden"

# ------------------------------------------------------------------- DEFECT
bad = golden_grid_sample(inp_nhwc, prepared)                   # exactly what a user gets today

print("\nDEFECT — the SAME golden, handed the precomputed grid it cannot recognise")
print("  golden(prepared_grid)     shape", tuple(bad.shape), " <-- 3x the width the device returns")
print("  device                    shape", tuple(dev.shape))
assert bad.shape != dev.shape, "shape must diverge"

print("\n  k_grid the golden derives from last_dim=6 :", prepared.shape[-1] // 2, "(true K is 1)")
print("  first precomputed point, read as floats   :", [float(v) for v in prepared[0, 0, 0]])
print("  ...the golden reads pair 0 as (x,y) = (%.3g, %.3g)  <-- these are int16 bit patterns, not coords"
      % (float(prepared[0, 0, 0, 0]), float(prepared[0, 0, 0, 1])))

n_cmp = min(bad.shape[2], dev.shape[2])
lhs, rhs = bad[:, :, :n_cmp, :], dev[:, :, :n_cmp, :]
print("\n  on the overlapping %d columns:" % n_cmp)
print("    golden :", [round(float(v), 3) for v in lhs.flatten()[:8]])
print("    device :", [round(float(v), 3) for v in rhs.flatten()[:8]])
nan_g = int(torch.isnan(lhs).sum()); nan_d = int(torch.isnan(rhs).sum())
finite = ~torch.isnan(lhs) & ~torch.isnan(rhs)
print("    NaNs   : golden %d of %d elements, device %d" % (nan_g, lhs.numel(), nan_d))
print("    max abs diff over elements finite in both:",
      float((lhs[finite] - rhs[finite]).abs().max()),
      " (input values span %g..%g)" % (float(inp_nhwc.min()), float(inp_nhwc.max())))
assert nan_g > 0 and nan_d == 0, "golden must emit NaN where the device does not"
assert float((lhs[finite] - rhs[finite]).abs().max()) > 1.0, "finite values must diverge too"

print("\n  golden(prepared, use_precomputed_grid=True) is byte-identical to")
print("  golden(prepared) with the flag omitted     :",
      torch.allclose(golden_grid_sample(inp_nhwc, prepared, use_precomputed_grid=True),
                     golden_grid_sample(inp_nhwc, prepared), equal_nan=True),
      " <-- the flag is swallowed by **_")

print("\nALL ASSERTIONS PASSED — the defect reproduces.")

```

</details>

### Suggested fix

Give the golden the parameter and the decode:

```python
def golden_grid_sample(input_tensor, grid, mode="bilinear", padding_mode="zeros",
                       align_corners=False, use_precomputed_grid=False,
                       batch_output_channels=False, grid_batching_factor=None, **_):
    if use_precomputed_grid:
        # 6 elements per point for bilinear, 2 for nearest - not 2 coordinates
        K = grid.shape[-1] // (2 if mode == "nearest" else 6)
        ...  # bit_cast the int16 corner indices, apply the four bf16 weights
```

The arithmetic is a weighted sum of four corners with invalid corners masked (`grid_sample_reader_common.hpp:347-360`), so it is a handful of lines. `nearest` needs the same treatment: its precomputed layout is 2 elements with a `-1` sentinel (`grid_sample_prepare_grid.cpp:89-96`), which the current golden would also read as a coordinate pair.

If decoding it in the golden is not wanted, the minimum is to **raise** on `use_precomputed_grid=True` instead of returning NaN, so the mismatch is loud.

Happy to send the PR either way — say which shape you want and I will put it up.

**Comment by p32929** [comment]
PR up: #57150. `+77 / -11`, one file.

Adds `use_precomputed_grid` to the golden and a reference that mirrors the device reader: `int16` corner indices back out of the bfloat16 payload, four bfloat16 weights, invalid corners masked, accumulate.

Verified device-free against the transcribed reader across three shapes with coordinates deliberately spilling outside `[-1, 1]` — **`max|diff| = 0` and zero NaNs on all three**, i.e. exact rather than within-tolerance, because the golden now consumes the same weights the device does. Standard mode is byte-for-byte the old code path, moved into an `else`.

One honest caveat, repeated from the PR: the `"nearest"` branch is written from the layout in `grid_sample_prepare_grid.cpp:89-96` and `PRECOMPUTED_GRID_ELEMENTS_PER_POINT_NEAREST`, reviewed by reading only — I measured bilinear, not nearest. Happy to cut that branch down to a `raise` and land the verified half first if you prefer.
